src/silentpayments.rs

```diff
@@ -0,0 +1,546 @@
+
+        let mut output33 = [0u8; 33];
+
+        let res = unsafe {
+            secp256k1_silentpayments_recipient_public_data_serialize(
+                secp.ctx().as_ptr(),
+                output33.as_mut_c_ptr(),
+                self.as_c_ptr(),
+            )
+        };
+
+        if res == 1 {
+            Ok(output33)
+        } else {
+            Err(SilentpaymentsPublicDataError::SerializationFailure)
+        }
+    }
+
+    /// Parse a 33-byte sequence into a silent_payments_public_data object.
+    pub fn parse<C: Verification>(secp: &Secp256k1<C>, input33: &[u8; 33]) -> Result<Self, SilentpaymentsPublicDataError> {
+
+        let empty_data = [0u8; constants::SILENT_PAYMENTS_PUBLIC_DATA_SIZE];
+        let mut silentpayments_public_data = ffi::SilentpaymentsPublicData::from_array(empty_data);
+
+        let res = unsafe {
+            secp256k1_silentpayments_recipient_public_data_parse(
+                secp.ctx().as_ptr(),
+                &mut silentpayments_public_data,
+                input33.as_c_ptr(),
+            )
+        };
+
+        if res == 1 {
+            let silentpayments_public_data = SilentpaymentsPublicData(silentpayments_public_data);
+            Ok(silentpayments_public_data)
+        } else {
+            Err(SilentpaymentsPublicDataError::ParseFailure)
+        }
+    }
+
+    /// Create Silent Payment shared secret.
+    pub fn recipient_create_shared_secret<C: Verification>(&self, secp: &Secp256k1<C>, recipient_scan_key: &SecretKey) -> Result<[u8; 33], SilentpaymentsPublicDataError> {
+        let mut output33 = [0u8; 33];
+
+        let res = unsafe {
+            secp256k1_silentpayments_recipient_create_shared_secret(
+                secp.ctx().as_ptr(),
+                output33.as_mut_c_ptr(),
+                recipient_scan_key.as_c_ptr(),
+                self.as_c_ptr(),
+            )
+        };
+
+        if res == 1 {
+            Ok(output33)
+        } else {
+            Err(SilentpaymentsPublicDataError::SharedSecretFailure)
+        }
+    }
+}
+
+/// Found outputs struct
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct SilentpaymentsFoundOutput(ffi::SilentpaymentsFoundOutput);
+
+impl CPtr for SilentpaymentsFoundOutput {
+    type Target = ffi::SilentpaymentsFoundOutput;
+
+    /// Obtains a const pointer suitable for use with FFI functions.
+    fn as_c_ptr(&self) -> *const Self::Target { &self.0 }
+
+    /// Obtains a mutable pointer suitable for use with FFI functions.
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target { &mut self.0 }
+}
+
+impl fmt::Display for SilentpaymentsFoundOutput {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+
+        let pubkey = XOnlyPublicKey::from(self.0.output.clone());
+
+        let buffer_str = pubkey.serialize()
+            .iter()
+            .map(|byte| format!("{:02x}", byte))
+            .collect::<String>();
+
+        write!(f, "{}", buffer_str)
+    }
+}
+
+impl SilentpaymentsFoundOutput {
```

Comment:
Maybe we should implement some getter functions too?

Reply:
What would be the use cases for the getter functions?

Reply:
I may be missing something, but in the tests we're just printing the `SilentpaymentsFoundOutput` returned by `silentpayments_recipient_scan_outputs`, and I can't figure out how I would take the x only pubkey to update my wallet state, let alone use the tweak to actually create the spending key for that output.

Reply:
I made a quick commit just to show you what I mean https://github.com/Sosthene00/rust-secp256k1/tree/bip352-add-getters